[JSC] Profile polymorphic callees in Wasm

Constellation opened the pull request:
#### 2db0579187cb8667dbc63e3e9b33480d340b512d
<pre>
[JSC] Profile polymorphic callees in Wasm
<a href="https://bugs.webkit.org/show_bug.cgi?id=301236">https://bugs.webkit.org/show_bug.cgi?id=301236</a>
<a href="https://rdar.apple.com/163153817">rdar://163153817</a>

Reviewed by Keith Miller.

This patch adds a mechanism collecting polymorphic callees in wasm
CallProfile. Previously, whenever we observe multiple callees, we mark
it as megamorphic. But this patch extends a mechanism to collect up to 3
callees. This is a preparation for more sophisticated inlining mechanism
for OMG which prioritizes frequently called one first for inlining.

1. We extend IPInt and BBQ to collect polymoprhic calles. For
   polymorphic callee cases, we use a shared thunk and this thunk counts
   which callee is called. And if we observe more than 3 callees, we
   mark it as megamorphic and stop using polymorphic callees.
   From collecting some information, we found 1, 2, or many callees are
   pretty common pattern. We are picking &quot;3&quot;, which is not wasting so
   much for unuseful cases, but still covering what we can see in many
   wasm code.
2. Collected callees &amp; call counts are stored in CallProfile::PolymorphicCallee.
   And MergedProfile read this concurrently to offer profiled
   information to OMG compiler.
3. In this patch, OMG is not leveraging this information so much yet. Only
   use of this information is sorting all call frequency and selecting the
   most frequently called one as an inlining target. In a subsequent patch,
   we will construct global InliningDecision data structure which views all
   call frequencies and decides which one should be inlined, and inlines
   multiple callees in one callsite when it is polymorphic callees and
   it has enough call counts for each case.

* JSTests/wasm/stress/wasm-call-indirect-megamorphic.js: Added.
(Initialize.table.with.the.eight.functions.async test.computeExpected):
(Initialize.table.with.the.eight.functions.async test):
* JSTests/wasm/stress/wasm-call-indirect-polymorphic.js: Added.
(Initialize.table.with.the.three.functions.async test):
* JSTests/wasm/stress/wasm-call-ref-megamorphic.js: Added.
(async test.computeExpected):
(async test):
* JSTests/wasm/stress/wasm-call-ref-polymorphic.js: Added.
(async test):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/NativeCallee.h:
(JSC::NativeCallee::category const): Deleted.
(JSC::NativeCallee::implementationVisibility const): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectCall):
* Source/JavaScriptCore/wasm/WasmCallProfile.cpp: Copied from Source/JavaScriptCore/wasm/WasmMergedProfile.cpp.
(JSC::Wasm::CallProfile::~CallProfile):
(JSC::Wasm::CallProfile::makePolymorphic):
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
(JSC::Wasm::CallProfile::observeCrossInstanceCall):
(JSC::Wasm::CallProfile::observeCallIndirect):
(JSC::Wasm::CallProfile::isMegamorphic):
(JSC::Wasm::CallProfile::monomorphic):
(JSC::Wasm::CallProfile::polymorphic):
(JSC::Wasm::CallProfile::makeMegamorphic):
* Source/JavaScriptCore/wasm/WasmMergedProfile.cpp:
(JSC::Wasm::MergedProfile::Candidates::markAsMegamorphic):
(JSC::Wasm::MergedProfile::Candidates::add):
(JSC::Wasm::MergedProfile::Candidates::merge):
(JSC::Wasm::MergedProfile::Candidates::finalize const):
(JSC::Wasm::MergedProfile::merge):
(JSC::Wasm::MergedProfile::CallSite::merge): Deleted.
* Source/JavaScriptCore/wasm/WasmMergedProfile.h:
(JSC::Wasm::MergedProfile::Candidates::isCalled const):
(JSC::Wasm::MergedProfile::Candidates::isEmpty const):
(JSC::Wasm::MergedProfile::Candidates::isMegamorphic const):
(JSC::Wasm::MergedProfile::Candidates::totalCount const):
(JSC::Wasm::MergedProfile::isCalled const):
(JSC::Wasm::MergedProfile::candidates const):
(JSC::Wasm::MergedProfile::CallSite::count const): Deleted.
(JSC::Wasm::MergedProfile::CallSite::callee const): Deleted.
(JSC::Wasm::MergedProfile::CallSite::isMegamorphic const): Deleted.
(JSC::Wasm::MergedProfile::callee const): Deleted.
(JSC::Wasm::MergedProfile::mutableSpan): Deleted.
(JSC::Wasm::MergedProfile::span const): Deleted.
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::createMergedProfile):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmThunks.cpp:
(JSC::Wasm::callPolymorphicCalleeGenerator):
* Source/JavaScriptCore/wasm/WasmThunks.h:

Canonical link: <a href="https://commits.webkit.org/301972@main">https://commits.webkit.org/301972@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67ee472698f77d38d4cb1628e1e3122be3a04396

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134888 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79174 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3e675ece-4c35-4037-bdc3-8653607688b3) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55795 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97146 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65059 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/14b09e63-c914-44f9-9ad1-244a5de58271) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130565 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38301 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77627 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3b67d5c4-77f6-45cc-b6bd-14b3a445c63f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32406 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78247 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119654 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32856 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137369 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126080 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41835 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105668 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110672 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105319 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50839 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29271 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51870 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19953 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54213 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60384 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159118 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53447 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39772 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56904 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55206 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmCallProfile.h:
```diff
@@ -57,22 +61,106 @@ class CallProfile {
+        if (boxedCallee & megamorphicCallee)
+            return nullptr;
+        if (boxedCallee & polymorphicCallee)
+            return nullptr;
```
could this be 
```
if (boxedCallee & calleeMask)
    return nullptr;
```
?

Constellation replied:
Changed.

tadeuzagallo commented on Source/JavaScriptCore/wasm/WasmCallProfile.h:
```diff
@@ -57,22 +61,106 @@ class CallProfile {
 private:
+    void makeMegamorphic()
+    {
+        m_boxedCallee = (m_boxedCallee | megamorphicCallee);
```
When you make it megamorphic, do you want to keep the polymorphic flag set? It looks like it's not used since `polymorphic()` bails if the megamorphic flag is set.

Constellation replied:
Yes, we would like to keep it. Since (1) this poly flag is used to destroy PolymorphicCallee in CallProfile destructor, and (2) we are intentionally not destroying it until CallProfile destructor timing as it can be concurrently read by the OMG compiler.

tadeuzagallo replied:
Oh, I totally missed the destructor, that's clever, nice!

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallProfile.h:
```diff
@@ -57,22 +61,106 @@ class CallProfile {
+    static constexpr EncodedJSValue initCallee = 0b0000;
+    static constexpr EncodedJSValue polymorphicCallee = 0b0100;
+    static constexpr EncodedJSValue megamorphicCallee = 0b1000;
+    static constexpr EncodedJSValue calleeMask = 0b1100;
```
Nit I would make these an 
```
enum State : EncodedJSValue {
    Monomorphic = 0,
    Polymorphic = 0b0100,
    Megamorphic = 0b1000
};
static constexpr EncodedJSValue calleeMask = 0b1100;
```

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmThunks.cpp:
```diff
@@ -165,6 +166,72 @@ MacroAssemblerCodeRef<JITThunkPtrTag> materializeBaselineDataGenerator(const Abs
+{
+    CCallHelpers jit;
+    JIT_COMMENT(jit, "callPolymorphicCalleeGenerator");
+
```
Probably worth a comment that we clobber the `nonPreservedNonArgumentGPR0` and `nonPreservedNonArgumentGPR1`

Constellation replied:
Added.

kmiller68 reviewed (approved):
r=me

Constellation committed:
[JSC] Profile polymorphic callees in Wasm
https://bugs.webkit.org/show_bug.cgi?id=301236
rdar://163153817

Reviewed by Keith Miller.

This patch adds a mechanism collecting polymorphic callees in wasm
CallProfile. Previously, whenever we observe multiple callees, we mark
it as megamorphic. But this patch extends a mechanism to collect up to 3
callees. This is a preparation for more sophisticated inlining mechanism
for OMG which prioritizes frequently called one first for inlining.

1. We extend IPInt and BBQ to collect polymoprhic calles. For
   polymorphic callee cases, we use a shared thunk and this thunk counts
   which callee is called. And if we observe more than 3 callees, we
   mark it as megamorphic and stop using polymorphic callees.
   From collecting some information, we found 1, 2, or many callees are
   pretty common pattern. We are picking "3", which is not wasting so
   much for unuseful cases, but still covering what we can see in many
   wasm code.
2. Collected callees & call counts are stored in CallProfile::PolymorphicCallee.
   And MergedProfile read this concurrently to offer profiled
   information to OMG compiler.
3. In this patch, OMG is not leveraging this information so much yet. Only
   use of this information is sorting all call frequency and selecting the
   most frequently called one as an inlining target. In a subsequent patch,
   we will construct global InliningDecision data structure which views all
   call frequencies and decides which one should be inlined, and inlines
   multiple callees in one callsite when it is polymorphic callees and
   it has enough call counts for each case.

* JSTests/wasm/stress/wasm-call-indirect-megamorphic.js: Added.
(Initialize.table.with.the.eight.functions.async test.computeExpected):
(Initialize.table.with.the.eight.functions.async test):
* JSTests/wasm/stress/wasm-call-indirect-polymorphic.js: Added.
(Initialize.table.with.the.three.functions.async test):
* JSTests/wasm/stress/wasm-call-ref-megamorphic.js: Added.
(async test.computeExpected):
(async test):
* JSTests/wasm/stress/wasm-call-ref-polymorphic.js: Added.
(async test):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/NativeCallee.h:
(JSC::NativeCallee::category const): Deleted.
(JSC::NativeCallee::implementationVisibility const): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitIndirectCall):
* Source/JavaScriptCore/wasm/WasmCallProfile.cpp: Copied from Source/JavaScriptCore/wasm/WasmMergedProfile.cpp.
(JSC::Wasm::CallProfile::~CallProfile):
(JSC::Wasm::CallProfile::makePolymorphic):
* Source/JavaScriptCore/wasm/WasmCallProfile.h:
(JSC::Wasm::CallProfile::observeCrossInstanceCall):
(JSC::Wasm::CallProfile::observeCallIndirect):
(JSC::Wasm::CallProfile::isMegamorphic):
(JSC::Wasm::CallProfile::monomorphic):
(JSC::Wasm::CallProfile::polymorphic):
(JSC::Wasm::CallProfile::makeMegamorphic):
* Source/JavaScriptCore/wasm/WasmMergedProfile.cpp:
(JSC::Wasm::MergedProfile::Candidates::markAsMegamorphic):
(JSC::Wasm::MergedProfile::Candidates::add):
(JSC::Wasm::MergedProfile::Candidates::merge):
(JSC::Wasm::MergedProfile::Candidates::finalize const):
(JSC::Wasm::MergedProfile::merge):
(JSC::Wasm::MergedProfile::CallSite::merge): Deleted.
* Source/JavaScriptCore/wasm/WasmMergedProfile.h:
(JSC::Wasm::MergedProfile::Candidates::isCalled const):
(JSC::Wasm::MergedProfile::Candidates::isEmpty const):
(JSC::Wasm::MergedProfile::Candidates::isMegamorphic const):
(JSC::Wasm::MergedProfile::Candidates::totalCount const):
(JSC::Wasm::MergedProfile::isCalled const):
(JSC::Wasm::MergedProfile::candidates const):
(JSC::Wasm::MergedProfile::CallSite::count const): Deleted.
(JSC::Wasm::MergedProfile::CallSite::callee const): Deleted.
(JSC::Wasm::MergedProfile::CallSite::isMegamorphic const): Deleted.
(JSC::Wasm::MergedProfile::callee const): Deleted.
(JSC::Wasm::MergedProfile::mutableSpan): Deleted.
(JSC::Wasm::MergedProfile::span const): Deleted.
* Source/JavaScriptCore/wasm/WasmModule.cpp:
(JSC::Wasm::Module::createMergedProfile):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmThunks.cpp:
(JSC::Wasm::callPolymorphicCalleeGenerator):
* Source/JavaScriptCore/wasm/WasmThunks.h:

Canonical link: https://commits.webkit.org/301972@main
